fairscale/utils/parallel.py

```diff
@@ -58,7 +58,7 @@ def enable_pytorch_sync_bn(module: torch.nn.Module) -> None:
             layer._specify_ddp_gpu_num(1)  # type: ignore
 
 
-def get_process_group_cached(ranks: Optional[Sequence[int]] = None) -> ProcessGroup:
+def get_process_group_cached(ranks: Optional[Sequence[int]] = None) -> Dict[str, ProcessGroup]:
```

Comment:
I suggest that we don't make this function be aware of specifics of all_gather or reduce_scatter. It should be able to support general and arbitrary keys.

In other words, I think the cache should change from using only `rank` as the key to a tuple `(name, rank)` as the key. This way, the caller just supply a `name` and a list of ranks and then, this will return either a new group or a cached group without caring about the actual value of the key.